juju/juju
# [JUJU-3666] Backport cmr secrets from 3.2 branch

Backport cmr secrets so it is available in the 2 year supported version of Juju.

This involves copying the 3.2 code for the secret facades, the remote relation workers, and some state code to persist the cross model consumer info:
- add new cross model secrets facades
- update secrets manager facade
- new cross model secrets watchers

The current v1 secrets manager facade is retained for compatibility with older 3.1 agents. A new v2 facade is added to handle the new cmr requirements.

## QA steps

run secrets ci tests (known permission error in removing an app owned secret)

plus

bootstrap a 3.1.2 controller
add vault as a backend
add a model and deploy juju-qa-dummy-source and juju-qa-dummy-sink and relate
set vault as the secret backend
create a secret and grant access
ensure secret can be accessed

upgrade the controller to this pr
check that the 3.1.2 model still works for secrets
access the secret
update the secret and check change hook firing

bootstrap a controller off this PR
add vault as a backend
deploy the above charms in a cross model scenario
check that secret access works, secret update workflow works, hooks fired
also check the single model scenario again
